test/exe/main_common_test.cc

```diff
@@ -226,32 +149,187 @@ TEST_F(MainCommonTest, LegacyMain) {
   EXPECT_EQ(EXIT_SUCCESS, return_code);
 }
 
-TEST_F(MainCommonTest, AdminRequestGetStatsAndQuit) {
+class AdminRequestTest : public MainCommonTest {
+protected:
+  AdminRequestTest()
+      : envoy_return_(false), envoy_started_(false), envoy_finished_(false),
+        pause_before_run_(false), pause_after_run_(false) {
+    addArg("--disable-hot-restart");
+  }
+
+  // Runs a an admin request specified in path, blocking until completion, and
+  // returning the response body.
+  std::string adminRequest(absl::string_view path, absl::string_view method) {
+    absl::Notification done;
+    std::string out;
+    main_common_->adminRequest(
+        path, method,
+        [&done, &out](const Http::HeaderMap& /*response_headers*/, absl::string_view body) {
+          out = std::string(body);
+          done.Notify();
+        });
+    done.WaitForNotification();
+    return out;
+  }
+
+  // Initiates Envoy running in its own thread.
+  void startEnvoy() {
+    envoy_thread_ = std::make_unique<Thread::Thread>([this]() {
+      // Note: main_common_ is accesesed in the testing thread, but
+      // is race-free, as MainCommon::run() does not return until
+      // triggered with an adminRequest POST to /quitquitquit, which
+      // is done in the testing thread.
+      main_common_ = std::make_unique<MainCommon>(argc(), argv());
+      envoy_started_ = true;
+      started_.Notify();
+      pauseResumeInterlock(pause_before_run_);
+      bool status = main_common_->run();
+      pauseResumeInterlock(pause_after_run_);
+      main_common_.reset();
+      envoy_finished_ = true;
+      envoy_return_ = status;
+      finished_.Notify();
+    });
+  }
+
+  // Conditionally pauses at a critical point in the Envoy thread waiting, waiting
+  // the for the test thread to try something. The test thread can then call resume_.Notify()
```

Comment:
Extra `the` snuck in. And I can't tell if the repetition of waiting is intention. If it is, SGTM :)

Reply:
done